### PR TITLE
add a panic example to std::from_utf8_unchecked

### PR DESCRIPTION
**Pull request by matthew-piziak**
Show that passing invalid bytes to `str::from_utf8_unchecked` is a runtime panic.

r? @steveklabnik

### CODE REVIEW

File: src/libcore/str/mod.rs [hunk 1]
```diff
@@ -299,6 +299,20 @@ unsafe fn from_raw_parts_mut<'a>(p: *mut u8, len: usize) -> &'a mut str {
 ///
 /// assert_eq!("💖", sparkle_heart);
 /// ```
+///
+/// Incorrect bytes:
+///
+/// ```
```
Comment (frewsxcv) [inline src/libcore/str/mod.rs]: This should probably be `no_run`, right?

Comment (frewsxcv) [inline src/libcore/str/mod.rs]: Actually, this example won't panic, so I guess it won't.

Comment (frewsxcv) [inline src/libcore/str/mod.rs]: Actually it might? The `from_utf8_unchecked` won't fail, but printing it might? Not sure.

Comment (matthew-piziak) [inline src/libcore/str/mod.rs]: This example won't panic?

Comment (frewsxcv) [inline src/libcore/str/mod.rs]: From what I understand, this function should never panic (hence it being "unchecked").

Comment (matthew-piziak) [inline src/libcore/str/mod.rs]: `unchecked ` means it won't return an `Err`. It won't panic either, but the call to `println!` will, as it attempts to print an invalid string. I'll split the lines out to make that clearer.

Comment (matthew-piziak) [inline src/libcore/str/mod.rs]: Whoops, never mind. It won't panic there either.

**Comment by frewsxcv** [comment]
While I can't think of any specific examples right now, it'd be great if we could somehow demonstrate memory safety issues, as described in the "Safety" section of this method.

**Comment by matthew-piziak** [comment]
@frewsxcv Yeah, I can't think of any examples either. If somebody reading this knows of one, can you let me know? Thanks in advance!

**Comment by matthew-piziak** [comment]
@ubsan ping!

**Comment by ubsan** [comment]
@matthew-piziak There are no memory safety issues with the current runtime. Someone could write something assuming valid utf-8, however, and read right off the end, however, for example with a `Chars` iterator. This is... not the most unsafe "undefined behavior".

**Comment by matthew-piziak** [comment]
@ubsan Gotcha, thanks. What do you think of this PR as-is, then?

File: src/libcore/str/mod.rs [hunk 2]
```diff
@@ -299,6 +299,21 @@ unsafe fn from_raw_parts_mut<'a>(p: *mut u8, len: usize) -> &'a mut str {
...
 /// assert_eq!("💖", sparkle_heart);
 /// ```
+///
+/// Incorrect bytes:
+///
+/// ```no_run
+/// use std::str;
+///
+/// // some invalid bytes, in a vector
+/// let sparkle_heart = vec![0, 159, 146, 150];
```
Comment (steveklabnik) [inline src/libcore/str/mod.rs]: maybe change this name, since it's no longer a sparkle heart?

File: src/libcore/str/mod.rs [hunk 3]
```diff
@@ -299,6 +299,21 @@ unsafe fn from_raw_parts_mut<'a>(p: *mut u8, len: usize) -> &'a mut str {
...
+///
+/// unsafe {
+///     let invalid_str = str::from_utf8_unchecked(&sparkle_heart);
+///     // prints undefined garbage to stdout
```
Comment (steveklabnik) [inline src/libcore/str/mod.rs]: "unknown" might be better than "undefined"

File: src/libcore/str/mod.rs (hunk 2 above)
Comment (matthew-piziak) [inline src/libcore/str/mod.rs]: Done. I also renamed the example in the checked conversion.

**Comment by ubsan** [comment]
@matthew-piziak I don't generally recommend committing UB in documentation, no matter how frivolous I may think that UB is :P.

I'm not sure... You could say "poses a safety issue" as well.

**Comment by matthew-piziak** [comment]
@ubsan Even with `no_run` braces?

**Comment by ubsan** [comment]
@matthew-piziak I meant more of "don't show people UB in documentation". I'd rather use correct examples.

**Comment by matthew-piziak** [comment]
@ubsan That's fair. Shall we close this PR?

**Comment by steveklabnik** [comment]
Yes, I agree. Thanks @matthew-piziak !

**Comment by matthew-piziak** [comment]
Decided not to show undefined behavior in documentation.

**Pull request closed without merging**